llama-index-core/llama_index/core/memory/vector_memory.py

```diff
@@ -0,0 +1,234 @@
+        # NOTE: we can't import VectorStoreIndex directly due to circular imports,
+        # which is why the type is Any
+        from llama_index.core.indices.vector_store import VectorStoreIndex
+
+        if not isinstance(value, VectorStoreIndex):
+            raise ValueError(
+                f"Expected 'vector_index' to be an instance of VectorStoreIndex, got {type(value)}"
+            )
+        return value
+
+    @classmethod
+    def class_name(cls) -> str:
+        """Get class name."""
+        return "VectorMemory"
+
+    @classmethod
+    def from_defaults(
+        cls,
+        vector_store: Optional[VectorStore] = None,
+        embed_model: Optional[EmbedType] = None,
+        index_kwargs: Optional[Dict] = None,
+        retriever_kwargs: Optional[Dict] = None,
+    ) -> "VectorMemory":
+        """Create vector memory.
+
+        Args:
+            vector_store (Optional[VectorStore]): vector store
+            embed_model (Optional[EmbedType]): embedding model
+            index_kwargs (Optional[Dict]): kwargs for initializing the index
+            retriever_kwargs (Optional[Dict]): kwargs for initializing the retriever
+
+        """
+        from llama_index.core.indices.vector_store import VectorStoreIndex
+
+        index_kwargs = index_kwargs or {}
+        retriever_kwargs = retriever_kwargs or {}
+
+        if vector_store is None:
+            # initialize a blank in-memory vector store
+            # NOTE: can't easily do that from `from_vector_store` at the moment.
+            index = VectorStoreIndex.from_documents(
+                [], embed_model=embed_model, **index_kwargs
+            )
+        else:
+            index = VectorStoreIndex.from_vector_store(
+                vector_store, embed_model=embed_model, **index_kwargs
+            )
+        return cls(vector_index=index, retriever_kwargs=retriever_kwargs)
+
+    def get(
+        self, input: Optional[str] = None, initial_token_count: int = 0, **kwargs: Any
+    ) -> List[ChatMessage]:
+        """Get chat history."""
+        if input is None:
+            raise ValueError("Input must be provided to get chat history.")
+
+        # retrieve from index
+        retriever = self.vector_index.as_retriever(**self.retriever_kwargs)
+        nodes = retriever.retrieve(input or "")
+
+        # retrieve underlying messages
+        messages = [
+            ChatMessage.parse_obj(sub_dict)
+            for node in nodes
+            for sub_dict in node.metadata["sub_dicts"]
+        ]
+
+        # add system message
+        if self.system_message:
+            messages = [
+                ChatMessage.from_str(self.system_message, role=MessageRole.SYSTEM)
+            ].extend(messages)
+
+        if self.return_single_message:
+            # condense all messages into a single message
+            messages = [
+                ChatMessage.from_str(
+                    " ".join([m.content for m in messages]), role=MessageRole.USER
```

Comment:
I wonder what this ends up looking like for chat histories that include tool calling (like openai tool calls)

Reply:
Hmm, I wonder if this could be from "SYSTEM" instead? The system is giving additional context to the current dialogue, which I think is more appropriate?

Maybe it has no difference, but with system we're saying something like: "Here's previous dialogue including tool calls and their results, use this information
in your response".

At least naturally this is something that the system would say versus the human user?

Reply:
The problem with making this a system message is similar to the issue above with inserting system messages.

And tbh, returning this statically as a user message might also be an issue, because we might end up in a place where the chat history has two user messages in a row (which also causes issues in some APIs 😓) -- I wonder If that should be handled at the LLM level or not though.